File: 133_Factorial Trailing Zeroes.py
"""
133. Factorial Trailing Zeroes
Count number of trailing zeros in n! by summing floor(n/5^i).
"""

def trailing_zeroes(n: int) -> int:
    count = 0
    while n > 0:
        n //= 5
        count += n
    return count

if __name__ == '__main__':
    print('trailingZeroes(3) =', trailing_zeroes(3))
    print('trailingZeroes(5) =', trailing_zeroes(5))
    print('trailingZeroes(100) =', trailing_zeroes(100))

# Time Complexity: O(log_5 n)
# Space Complexity: O(1)

File: 133_Factorial Trailing Zeroes.cpp
/*
133. Factorial Trailing Zeroes
Count trailing zeros in n! by counting factors of 5.
*/
#include <bits/stdc++.h>
using namespace std;

int trailingZeroes(int n) {
    int count = 0;
    while (n > 0) {
        n /= 5;
        count += n;
    }
    return count;
}

int main() {
    cout << "trailingZeroes(3) = " << trailingZeroes(3) << "\n"; // 0
    cout << "trailingZeroes(5) = " << trailingZeroes(5) << "\n"; // 1
    cout << "trailingZeroes(100) = " << trailingZeroes(100) << "\n"; // 24
    return 0;
}

// Time Complexity: O(log_5 n)
// Space Complexity: O(1)
